File: Module 4 (Complete project)/Backend/controllers/class.controller.js
const Class = require("../models/class.model");
const Timetable = require("../models/timetable.schema");

// Create Class and automatically create timetable
exports.createClass = async (req, res) => {
  let newClass;
  try {
    const { className, subjects } = req.body;

    if (!className || !subjects || !Array.isArray(subjects) || subjects.length === 0) {
      return res.status(400).json({ message: "Class name and at least one subject are required" });
    }

    // Check duplicate
    const existingClass = await Class.findOne({ className });
    if (existingClass) {
      return res.status(400).json({ message: "Class name already exists" });
    }

    // Create class
    newClass = new Class({ className, subjects });
    await newClass.save();

    // Create timetable (days & slots default to model)
    const newTimetable = new Timetable({
      classRef: newClass._id,
      className: newClass.className
    });
    await newTimetable.save();

    res.status(201).json({
      message: "Class and timetable created successfully",
      class: newClass,
      timetable: newTimetable
    });

  } catch (error) {
    console.error("Error creating class or timetable:", error);
    if (newClass) await Class.findByIdAndDelete(newClass._id);
    res.status(500).json({ message: "Something went wrong. Please try again." });
  }
};


exports.getclassesadmin=async (req, res) => {
  try {
    const {classId}=req.params;
    const classes = await Class.findById(classId).sort({ className: 1 });
    res.status(200).json(classes);
  } catch (error) {
    console.error("Error fetching classes:", error);
    res.status(500).json({ message: "Something went wrong. Please try again." });
  }
}

// Get all classes (for dropdown)
exports.getClasses = async (req, res) => {
  try {
    const classes = await Class.find().select("_id className").sort({ className: 1 });
    res.status(200).json(classes);
  } catch (error) {
    console.error("Error fetching classes:", error);
    res.status(500).json({ message: "Something went wrong. Please try again." });
  }
};

// Get timetable for a specific class
exports.getTimetable = async (req, res) => {
  try {
    console.log(req.params);
   const { classname } = req.params; // <- use params, not body
    const timetable = await Timetable.findOne({ className: classname });

    if (!timetable) {
      return res.status(404).json({ message: "Timetable not found" });
    }

    res.status(200).json({ timetable });
  } catch (error) {
    console.error("Error fetching timetable:", error);
    res.status(500).json({ message: "Something went wrong. Please try again." });
  }
};


// Create timetable manually (if it does not exist yet)
exports.createTimetable = async (req, res) => {
  try {
    const { classRef, className, days } = req.body;

    if (!classRef || !className) {
      return res.status(400).json({ message: "Class reference and name are required" });
    }

    // Prevent duplicate
    const existingTimetable = await Timetable.findOne({ classRef });
    if (existingTimetable) {
      return res.status(400).json({ message: "Timetable already exists for this class" });
    }

    const newTimetable = new Timetable({
      classRef,
      className,
      days: days || undefined
    });
    await newTimetable.save();

    res.status(201).json({
      message: "Timetable created successfully",
      timetable: newTimetable
    });
  } catch (error) {
    console.error("Error creating timetable:", error);
    res.status(500).json({ message: "Something went wrong. Please try again." });
  }
};

// Update existing timetable
exports.updateTimetable = async (req, res) => {
  try {
    const { classRef, days } = req.body;

    if (!classRef || !days || !Array.isArray(days) || days.length === 0) {
      return res.status(400).json({ message: "Class reference and days are required" });
    }

    const timetable = await Timetable.findOne({ classRef });

    if (!timetable) {
      return res.status(404).json({ message: "Timetable not found for this class" });
    }

    timetable.days = days;
    await timetable.save();

    res.status(200).json({
      message: "Timetable updated successfully",
      timetable
    });

  } catch (error) {
    console.error("Error updating timetable:", error);
    res.status(500).json({ message: "Something went wrong. Please try again." });
  }
};

File: Experiment 9 & 10  -Final Project/backend/src/main/java/com/example/backend/controller/ClassController.java
package com.example.backend.controller;

import com.example.backend.model.ClassModel;
import com.example.backend.model.Timetable;
import com.example.backend.repository.ClassRepository;
import com.example.backend.repository.TimetableRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.util.List;

@RestController
@RequestMapping("/api/class")
@CrossOrigin
public class ClassController {

    @Autowired
    private ClassRepository classRepository;

    @Autowired
    private TimetableRepository timetableRepository;

    // Create Class and automatically create timetable
    @PostMapping("/")
    public ResponseEntity<?> createClass(@RequestBody ClassModel classModel) {
        try {
            if (classModel.getClassName() == null || classModel.getSubjects() == null || classModel.getSubjects().isEmpty()) {
                return ResponseEntity.badRequest().body("Class name and at least one subject are required");
            }

            ClassModel existingClass = classRepository.findByClassName(classModel.getClassName());
            if (existingClass != null) {
                return ResponseEntity.badRequest().body("Class name already exists");
            }

            ClassModel newClass = classRepository.save(classModel);

            // Create default timetable
            Timetable timetable = new Timetable();
            timetable.setClassRef(newClass.getId());
            timetable.setClassName(newClass.getClassName());
            timetableRepository.save(timetable);

            return ResponseEntity.status(201).body(newClass);
        } catch (Exception e) {
            e.printStackTrace();
            return ResponseEntity.status(500).body("Something went wrong. Please try again.");
        }
    }

    // Get all classes
    @GetMapping("/get")
    public ResponseEntity<List<ClassModel>> getClasses() {
        try {
            List<ClassModel> classes = classRepository.findAll();
            return ResponseEntity.ok(classes);
        } catch (Exception e) {
            e.printStackTrace();
            return ResponseEntity.status(500).build();
        }
    }

    // Get timetable for a specific class
    @GetMapping("/timetable/{classname}")
    public ResponseEntity<?> getTimetable(@PathVariable String classname) {
        try {
            Timetable timetable = timetableRepository.findByClassName(classname).orElse(null);
            if (timetable == null) {
                return ResponseEntity.status(404).body("Timetable not found");
            }
            return ResponseEntity.ok(timetable);
        } catch (Exception e) {
            e.printStackTrace();
            return ResponseEntity.status(500).body("Something went wrong. Please try again.");
        }
    }

    // Create timetable manually
    @PostMapping("/createtimetable")
    public ResponseEntity<?> createTimetable(@RequestBody Timetable timetable) {
        try {
            if (timetable.getClassRef() == null || timetable.getClassName() == null) {
                return ResponseEntity.badRequest().body("Class reference and name are required");
            }

            Timetable existing = timetableRepository.findByClassRef(timetable.getClassRef()).orElse(null);
            if (existing != null) {
                return ResponseEntity.badRequest().body("Timetable already exists for this class");
            }

            Timetable savedTimetable = timetableRepository.save(timetable);
            return ResponseEntity.status(201).body(savedTimetable);
        } catch (Exception e) {
            e.printStackTrace();
            return ResponseEntity.status(500).body("Something went wrong. Please try again.");
        }
    }

    // Update existing timetable
    @PutMapping("/updatetimetable")
    public ResponseEntity<?> updateTimetable(@RequestBody Timetable timetable) {
        try {
            if (timetable.getClassRef() == null || timetable.getDays() == null || timetable.getDays().isEmpty()) {
                return ResponseEntity.badRequest().body("Class reference and days are required");
            }

            Timetable existing = timetableRepository.findByClassRef(timetable.getClassRef()).orElse(null);
            if (existing == null) {
                return ResponseEntity.status(404).body("Timetable not found for this class");
            }

            existing.setDays(timetable.getDays());
            timetableRepository.save(existing);

            return ResponseEntity.ok(existing);
        } catch (Exception e) {
            e.printStackTrace();
            return ResponseEntity.status(500).body("Something went wrong. Please try again.");
        }
    }

    // Get class details by admin
    @GetMapping("/admin/{classId}")
    public ResponseEntity<?> getClassAdmin(@PathVariable String classId) {
        try {
            ClassModel classModel = classRepository.findById(classId).orElse(null);
            if (classModel == null) {
                return ResponseEntity.status(404).body("Class not found");
            }
            return ResponseEntity.ok(classModel);
        } catch (Exception e) {
            e.printStackTrace();
            return ResponseEntity.status(500).body("Something went wrong. Please try again.");
        }
    }
}
